File: local-server/upload_server.py
#!/usr/bin/env python3
"""
BULLETPROOF Local Upload Server
"""

import os
import json
import hashlib
import time
import traceback
from datetime import datetime
from flask import Flask, request, jsonify
from werkzeug.utils import secure_filename

app = Flask(__name__)

UPLOAD_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "uploads")
CHUNKS_FOLDER = os.path.join(UPLOAD_FOLDER, "chunks")
SESSIONS_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "sessions.json")
LOG_FILE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "server.log")

os.makedirs(UPLOAD_FOLDER, exist_ok=True)
os.makedirs(CHUNKS_FOLDER, exist_ok=True)

def log(msg):
    ts = datetime.now().strftime("%H:%M:%S")
    line = "[%s] %s" % (ts, msg)
    print(line)
    with open(LOG_FILE, "a") as f:
        f.write(line + "\n")

def load_sessions():
    if os.path.exists(SESSIONS_FILE):
        try:
            with open(SESSIONS_FILE, "r") as f:
                return json.load(f)
        except Exception as e:
            log("ERROR loading sessions.json: %s" % e)
    return {}

def save_sessions(data):
    try:
        with open(SESSIONS_FILE, "w") as f:
            json.dump(data, f, indent=2)
    except Exception as e:
        log("ERROR saving sessions.json: %s" % e)

sessions = load_sessions()
chunk_registry = {}

@app.route("/api/v1/upload/init", methods=["POST"])
def init_upload():
    try:
        raw = request.get_data(as_text=True)
        data = request.get_json() or {}

        log("RAW INIT BODY: %s" % raw[:500])
        log("PARSED KEYS: %s" % list(data.keys()))

        task_id = data.get("taskId", "")
        file_name = data.get("fileName", "unknown")
        mime_type = data.get("mimeType", "application/octet-stream")
        total_bytes = data.get("totalBytes", 0)
        total_chunks = data.get("totalChunks", 0)
        checksum = data.get("checksum", "")

        if not task_id:
            task_id = "task_%s" % hashlib.md5(("%s%s" % (file_name, time.time())).encode()).hexdigest()[:12]
            log("GENERATED taskId: %s" % task_id)

        session_id = "sess_%s" % hashlib.md5(("%s%s" % (task_id, time.time())).encode()).hexdigest()[:16]

        sessions[task_id] = {
            "sessionId": session_id,
            "fileName": secure_filename(file_name),
            "mimeType": mime_type,
            "totalBytes": total_bytes,
            "totalChunks": total_chunks,
            "checksum": checksum,
            "uploadedChunks": [],
            "createdAt": time.time(),
            "status": "in_progress"
        }
        chunk_registry[task_id] = {}
        save_sessions(sessions)

        log("INIT SUCCESS: taskId=%s | sessionId=%s | file=%s | chunks=%s" % (task_id, session_id, file_name, total_chunks))

        return jsonify({
            "success": True,
            "sessionId": session_id,
            "uploadUrl": "http://%s/api/v1/upload/chunk" % request.host,
            "expiresAt": int((time.time() + 3600) * 1000),
            "taskId": task_id
        })
    except Exception as e:
        log("INIT CRASH: %s" % traceback.format_exc())
        return jsonify({"success": False, "message": str(e)}), 500

@app.route("/api/v1/upload/chunk", methods=["POST"])
def upload_chunk():
    try:
        task_id = request.form.get("task_id", "")
        chunk_index = request.form.get("chunk_index", "-1")
        total_chunks = request.form.get("total_chunks", "0")

        log("CHUNK REQUEST: task_id=%s | chunk_index=%s | total_chunks=%s" % (task_id, chunk_index, total_chunks))

        if not task_id or chunk_index == "-1":
            return jsonify({"success": False, "message": "Missing parameters"}), 400

        chunk_idx = int(chunk_index)
        total = int(total_chunks)

        # FALLBACK: auto-create session if missing
        if task_id not in sessions:
            log("CHUNK FALLBACK: auto-creating session for %s" % task_id)
            sessions[task_id] = {
                "sessionId": "sess_fallback_%s" % task_id[:8],
                "fileName": "unknown_%s.bin" % task_id[:8],
                "mimeType": "application/octet-stream",
                "totalBytes": 0,
                "totalChunks": total,
                "checksum": "",
                "uploadedChunks": [],
                "createdAt": time.time(),
                "status": "in_progress"
            }
            chunk_registry[task_id] = {}
            save_sessions(sessions)

        if "file" not in request.files:
            return jsonify({"success": False, "message": "No file part"}), 400

        file = request.files["file"]
        chunk_filename = "%s_chunk_%s" % (task_id, chunk_idx)
        chunk_path = os.path.join(CHUNKS_FOLDER, chunk_filename)
        file.save(chunk_path)

        if chunk_idx not in sessions[task_id]["uploadedChunks"]:
            sessions[task_id]["uploadedChunks"].append(chunk_idx)
            save_sessions(sessions)

        size = os.path.getsize(chunk_path)
        log("CHUNK STORED: taskId=%s | chunk=%s/%s | %s bytes" % (task_id, chunk_idx + 1, total, size))

        return jsonify({
            "success": True,
            "eTag": "etag_%s" % chunk_idx,
            "chunkIndex": chunk_idx,
            "message": "Chunk uploaded"
        })
    except Exception as e:
        log("CHUNK CRASH: %s" % traceback.format_exc())
        return jsonify({"success": False, "message": str(e)}), 500

@app.route("/api/v1/upload/commit", methods=["POST"])
def commit_upload():
    try:
        raw = request.get_data(as_text=True)
        data = request.get_json() or {}

        log("RAW COMMIT BODY: %s" % raw[:500])

        task_id = data.get("taskId") or data.get("task_id", "")
        file_name = data.get("fileName", "unknown")
        total_chunks = data.get("totalChunks", 0)

        log("COMMIT PARSED: taskId=%s | fileName=%s | totalChunks=%s" % (task_id, file_name, total_chunks))
        log("ALL SESSIONS: %s" % list(sessions.keys()))

        if not task_id:
            return jsonify({"success": False, "message": "Missing taskId"}), 400

        # FALLBACK: auto-create session if missing
        if task_id not in sessions:
            log("COMMIT FALLBACK: auto-creating session for %s" % task_id)
            sessions[task_id] = {
                "sessionId": "sess_fallback_%s" % task_id[:8],
                "fileName": secure_filename(file_name) if file_name else "unknown_%s.bin" % task_id[:8],
                "mimeType": "application/octet-stream",
                "totalBytes": 0,
                "totalChunks": total_chunks,
                "checksum": "",
                "uploadedChunks": list(range(total_chunks)),
                "createdAt": time.time(),
                "status": "in_progress"
            }
            save_sessions(sessions)

        # Reassemble
        final_filename = secure_filename(file_name)
        if not final_filename:
            final_filename = "upload_%s.bin" % task_id[:8]

        final_path = os.path.join(UPLOAD_FOLDER, final_filename)
        counter = 1
        original = final_path
        while os.path.exists(final_path):
            name, ext = os.path.splitext(original)
            final_path = "%s_%s%s" % (name, counter, ext)
            counter += 1

        with open(final_path, "wb") as outfile:
            for i in range(total_chunks):
                chunk_path = os.path.join(CHUNKS_FOLDER, "%s_chunk_%s" % (task_id, i))
                if os.path.exists(chunk_path):
                    with open(chunk_path, "rb") as infile:
                        outfile.write(infile.read())
                    os.remove(chunk_path)
                    log("  -> merged chunk %s" % i)
                else:
                    log("  -> MISSING chunk %s" % i)

        sessions[task_id]["status"] = "completed"
        sessions[task_id]["remoteUrl"] = "http://%s/uploads/%s" % (request.host, os.path.basename(final_path))
        sessions[task_id]["fileId"] = "file_%s" % task_id
        save_sessions(sessions)

        file_size = os.path.getsize(final_path)
        log("COMMIT SUCCESS: taskId=%s | saved=%s | %s bytes" % (task_id, os.path.basename(final_path), file_size))

        return jsonify({
            "success": True,
            "remoteUrl": sessions[task_id]["remoteUrl"],
            "fileId": sessions[task_id]["fileId"],
            "message": "Upload completed"
        })
    except Exception as e:
        log("COMMIT CRASH: %s" % traceback.format_exc())
        return jsonify({"success": False, "message": str(e)}), 500

@app.route("/api/v1/upload/status/<task_id>", methods=["GET"])
def upload_status(task_id):
    if task_id not in sessions:
        return jsonify({"success": False, "message": "Task not found"}), 404

    return jsonify({
        "taskId": task_id,
        "status": sessions[task_id]["status"],
        "uploadedChunks": sessions[task_id]["uploadedChunks"],
        "remoteUrl": sessions[task_id].get("remoteUrl")
    })

@app.route("/api/v1/upload/session/refresh", methods=["POST"])
def refresh_session():
    data = request.get_json() or {}
    task_id = data.get("taskId", "")

    if task_id not in sessions:
        return jsonify({"success": False, "message": "Session not found"}), 404

    new_session_id = "sess_%s" % hashlib.md5(("%s%s" % (task_id, time.time())).encode()).hexdigest()[:16]
    sessions[task_id]["sessionId"] = new_session_id
    save_sessions(sessions)

    return jsonify({
        "success": True,
        "sessionId": new_session_id,
        "uploadUrl": "http://%s/api/v1/upload/chunk" % request.host,
        "expiresAt": int((time.time() + 3600) * 1000)
    })

@app.route("/uploads/<filename>", methods=["GET"])
def serve_file(filename):
    return app.send_from_directory(UPLOAD_FOLDER, filename)

@app.route("/api/v1/health", methods=["GET"])
def health_check():
    return jsonify({
        "status": "ok",
        "activeSessions": len(sessions),
        "uploadFolder": UPLOAD_FOLDER
    })

@app.route("/", methods=["GET"])
def index():
    completed = [
        {"taskId": k, "fileName": v["fileName"], "fileId": v.get("fileId"), "remoteUrl": v.get("remoteUrl")}
        for k, v in sessions.items() if v.get("status") == "completed"
    ]
    return jsonify({
        "server": "Android Upload SDK Local Server",
        "version": "1.1.0",
        "uploadFolder": UPLOAD_FOLDER,
        "completedUploads": len(completed),
        "uploads": completed
    })

if __name__ == "__main__":
    print("=" * 60)
    print("BULLETPROOF Upload Server")
    print("=" * 60)
    print("Upload folder: %s" % UPLOAD_FOLDER)
    print("Sessions file: %s" % SESSIONS_FILE)
    print("Log file: %s" % LOG_FILE)
    print("API Base URL: http://localhost:5000/api/v1/")
    print("")
    print("For Android Emulator: use http://10.0.2.2:5000/api/v1/")
    print("For Physical Device: use http://YOUR_MAC_IP:5000/api/v1/")
    print("=" * 60)

    app.run(host="0.0.0.0", port=5000, debug=False, threaded=True)

File: local-server/upload_server.js
const express = require("express");
const multer = require("multer");
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const app = express();
app.use(express.json());

const UPLOAD_FOLDER = path.join(__dirname, "uploads");
const CHUNKS_FOLDER = path.join(UPLOAD_FOLDER, "chunks");

// Ensure folders exist
[UPLOAD_FOLDER, CHUNKS_FOLDER].forEach((dir) => {
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
});

// In-memory storage
const sessions = {};
const chunkRegistry = {};

// Configure multer for chunk uploads
const upload = multer({ dest: CHUNKS_FOLDER });

// Helper: Generate IDs
function generateId(prefix) {
  return prefix + "_" + crypto.randomBytes(8).toString("hex");
}

// Helper: Get timestamp
function now() {
  return new Date().toLocaleTimeString();
}

// 1. Initialize Upload
app.post("/api/v1/upload/init", (req, res) => {
  const { fileName, mimeType, totalBytes, totalChunks, checksum } = req.body;

  const sessionId = generateId("sess");
  const taskId = generateId("task");

  sessions[taskId] = {
    sessionId,
    fileName: path.basename(fileName),
    mimeType,
    totalBytes,
    totalChunks,
    checksum,
    uploadedChunks: [],
    createdAt: Date.now(),
    status: "in_progress",
  };
  chunkRegistry[taskId] = {};

  console.log(`[${now()}] INIT: ${taskId} | ${fileName} | ${totalBytes} bytes`);

  res.json({
    success: true,
    sessionId,
    uploadUrl: `http://${req.headers.host}/api/v1/upload/chunk`,
    expiresAt: Date.now() + 3600000,
    taskId,
  });
});

// 2. Upload Chunk
app.post("/api/v1/upload/chunk", upload.single("file"), (req, res) => {
  const { task_id, chunk_index, total_chunks, session_id } = req.body;

  if (!task_id || chunk_index === undefined) {
    return res.status(400).json({ success: false, message: "Missing parameters" });
  }

  const chunkIdx = parseInt(chunk_index);
  const chunkFileName = `${task_id}_chunk_${chunkIdx}`;
  const chunkPath = path.join(CHUNKS_FOLDER, chunkFileName);

  // Move uploaded file to proper chunk name
  if (req.file) {
    fs.renameSync(req.file.path, chunkPath);
  }

  // Register chunk
  if (!chunkRegistry[task_id]) chunkRegistry[task_id] = {};
  chunkRegistry[task_id][chunkIdx] = `etag_${chunkIdx}`;

  if (sessions[task_id]) {
    if (!sessions[task_id].uploadedChunks.includes(chunkIdx)) {
      sessions[task_id].uploadedChunks.push(chunkIdx);
    }
  }

  const size = fs.existsSync(chunkPath) ? fs.statSync(chunkPath).size : 0;
  console.log(`[${now()}] CHUNK: ${task_id} | chunk ${parseInt(chunk_index) + 1}/${total_chunks} | ${size} bytes`);

  res.json({
    success: true,
    eTag: `etag_${chunkIdx}`,
    chunkIndex: chunkIdx,
    message: "Chunk uploaded",
  });
});

// 3. Commit Upload
app.post("/api/v1/upload/commit", (req, res) => {
  const { taskId, sessionId, fileName, totalChunks } = req.body;

  if (!sessions[taskId]) {
    return res.status(404).json({ success: false, message: "Session not found" });
  }

  const finalName = path.basename(fileName);
  let finalPath = path.join(UPLOAD_FOLDER, finalName);

  // Handle duplicates
  let counter = 1;
  const originalPath = finalPath;
  while (fs.existsSync(finalPath)) {
    const ext = path.extname(originalPath);
    const base = path.basename(originalPath, ext);
    finalPath = path.join(UPLOAD_FOLDER, `${base}_${counter}${ext}`);
    counter++;
  }

  // Reassemble chunks
  const writeStream = fs.createWriteStream(finalPath);
  for (let i = 0; i < totalChunks; i++) {
    const chunkPath = path.join(CHUNKS_FOLDER, `${taskId}_chunk_${i}`);
    if (fs.existsSync(chunkPath)) {
      const data = fs.readFileSync(chunkPath);
      writeStream.write(data);
      fs.unlinkSync(chunkPath); // Clean up
    }
  }
  writeStream.end();

  sessions[taskId].status = "completed";
  sessions[taskId].remoteUrl = `http://${req.headers.host}/uploads/${path.basename(finalPath)}`;
  sessions[taskId].fileId = `file_${taskId}`;

  const stats = fs.statSync(finalPath);
  console.log(`[${now()}] COMMIT: ${taskId} | ${path.basename(finalPath)} | ${stats.size} bytes | SAVED`);

  res.json({
    success: true,
    remoteUrl: sessions[taskId].remoteUrl,
    fileId: sessions[taskId].fileId,
    message: "Upload completed",
  });
});

// 4. Check Status
app.get("/api/v1/upload/status/:taskId", (req, res) => {
  const { taskId } = req.params;

  if (!sessions[taskId]) {
    return res.status(404).json({ success: false, message: "Task not found" });
  }

  res.json({
    taskId,
    status: sessions[taskId].status,
    uploadedChunks: sessions[taskId].uploadedChunks,
    remoteUrl: sessions[taskId].remoteUrl || null,
  });
});

// 5. Refresh Session
app.post("/api/v1/upload/session/refresh", (req, res) => {
  const { sessionId, taskId } = req.body;

  if (!sessions[taskId]) {
    return res.status(404).json({ success: false, message: "Session not found" });
  }

  const newSessionId = generateId("sess");
  sessions[taskId].sessionId = newSessionId;

  console.log(`[${now()}] REFRESH: ${taskId} | new session ${newSessionId}`);

  res.json({
    success: true,
    sessionId: newSessionId,
    uploadUrl: `http://${req.headers.host}/api/v1/upload/chunk`,
    expiresAt: Date.now() + 3600000,
  });
});

// Serve uploaded files
app.get("/uploads/:filename", (req, res) => {
  const filePath = path.join(UPLOAD_FOLDER, req.params.filename);
  if (fs.existsSync(filePath)) {
    res.sendFile(filePath);
  } else {
    res.status(404).json({ error: "File not found" });
  }
});

// Health check
app.get("/api/v1/health", (req, res) => {
  res.json({
    status: "ok",
    activeSessions: Object.keys(sessions).length,
    uploadFolder: UPLOAD_FOLDER,
  });
});

// Info page
app.get("/", (req, res) => {
  const uploads = Object.entries(sessions)
    .filter(([_, s]) => s.status === "completed")
    .map(([taskId, s]) => ({
      taskId,
      fileName: s.fileName,
      fileId: s.fileId,
      remoteUrl: s.remoteUrl,
    }));

  res.json({
    server: "Android Upload SDK Local Server (Node.js)",
    version: "1.0.0",
    uploadFolder: UPLOAD_FOLDER,
    completedUploads: uploads.length,
    uploads,
  });
});

const PORT = process.env.PORT || 5000;
app.listen(PORT, "0.0.0.0", () => {
  console.log("=".repeat(60));
  console.log("Android Upload SDK - Local Server (Node.js)");
  console.log("=".repeat(60));
  console.log(`Upload folder: ${UPLOAD_FOLDER}`);
  console.log(`API Base URL: http://localhost:${PORT}/api/v1/`);
  console.log("");
  console.log(`For Android Emulator: use http://10.0.2.2:${PORT}/api/v1/`);
  console.log(`For Physical Device: use http://YOUR_MAC_IP:${PORT}/api/v1/`);
  console.log("=".repeat(60));
});
